Default Value Support and Additional envVars for TagOnly

h1dden-da3m0n opened the pull request:
## Summary
This PR is an alternative to [image-tag-parameter-plugin#4](https://github.com/jenkinsci/image-tag-parameter-plugin/pull/4).
It adds the support for the user to set a default tag value that should get preselected if it exists in the fetched tags list.
Additionally, this PR adds the support do directly access the selected tag without the image name by exposing it as an additional environment variable (see readme changes for how to access this).

## Tests
Jenkins 2.222.1 and 2.235.1 (aka LTS)
additionally checked support for [rebuild-plugin](https://github.com/jenkinsci/rebuild-plugin) support.

---

Should there be anything that prevents this from a merge let me know and ill try to fix it ASAP.

And finally, a big thank you to @Reuuke for allowing me to use some of their original PR! :beers:

h1dden-da3m0n commented:
A bit off-topic, but please allow me to ask:
Why does this repo not use Release Drafter like some of the other newer JenkinsCI repositories?

(I am just curious xD)

maduma commented on src/main/java/io/jenkins/plugins/luxair/ImageTag.java:
```diff
@@ -48,9 +48,9 @@
             rtn[0] = m.group(1);
             rtn[1] = m.group(2);
-            logger.info("realm:" + rtn[0] + ": service:" + rtn[1] + ":");
+            logger.info(() -> "realm:" + rtn[0] + ": service:" + rtn[1] + ":");
```
Why using lambda here ? Keep it simple

h1dden-da3m0n replied:
Sorry, this was SonarLint messing with me. I'll fix this and its repetitions

maduma commented on src/main/java/io/jenkins/plugins/luxair/ImageTag.java:
```diff
@@ -1,32 +1,32 @@
         List<String> tags = getImageTagsFromRegistry(image, registry, token);
         return tags.stream().filter(tag -> tag.matches(filter))
-            .map(tag -> image + ":" + tag)
+            .sorted()
```
Please do not sort. I want to have the last pushed tag in first in the selection box, or better sorted using maven versioning (e.g. v2.1.2, v2.1.0, v1.12.4, v1.7.23, v0.2.4)

h1dden-da3m0n replied:
Hmmm okay, sure can do. 
Would it be okay with you if the sort order where configurable? 
Or should I open another PR for something like that?

maduma replied:
Let split the feature (configurable order) in a other PR.

h1dden-da3m0n replied:
due to the nature of tags allowing free form strings (e.g.: latest or lts) make it not really straight forward two compare them properly (even things like v1.0-alpine would be problematic).

maduma commented on src/main/java/io/jenkins/plugins/luxair/ImageTag.java:
```diff
@@ -48,9 +48,9 @@
+            logger.info(() -> "realm:" + rtn[0] + ": service:" + rtn[1] + ":");
         } else {
-            logger.warning("No AuthService available from " + url);
+            logger.warning(() -> "No AuthService available from " + url);
```
Why using lambda here ? Keep it simple

maduma commented on src/main/java/io/jenkins/plugins/luxair/ImageTagParameterDefinition.java:
```diff
@@ -89,30 +98,39 @@ private StandardUsernamePasswordCredentials findCredential(String credentialId)
                 }
             }
-            logger.warning("Cannot find credential for :" + credentialId + ":");
+            logger.warning(() -> "Cannot find credential for :" + credentialId + ":");
```
Why using lambda here ? Keep it simple

h1dden-da3m0n commented:
Okay addressed all comments as far as I am concern.
Sadly I the version/tag ordering/sorting is something I have no real plan of attack for YET, so I reverted to the original behaviour `Collections.sort(tags, Collections.reverseOrder());` but doing so inside of the returned stream. 

I hope this is sufficient for now, I'll be playing around with ideas about how to handle tag sorting properly and open a PR once I have something to present.

h1dden-da3m0n commented:
Okay, I might have an idea, even though it might change the behaviour for images with tags not conforming with versions.
There is `hudson.util.VersionNumber` class that is comparable and will work for anything semantically versioned or with maven versioning, but tags like latest or LTS will not be handled.
Personally, I think that is fine since I discourage the use of these "shifting" tags in builds, but I would like your opinion on that first before I make this change using this class.

h1dden-da3m0n commented:
@maduma Thy for the merge :grin:

maduma commented:
hudson.util.VersionNumber is good. And yes using 'shifting' tags is not best practice.
This should be an option, disabled by default (default is reverse natural sort)
Newest version of should display first in the list
